ext/libp2p/lib.rs

```diff
@@ -0,0 +1,93 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+use deno_core::anyhow::{anyhow, Context, Result};
+use deno_core::error::AnyError;
+use deno_core::{include_js_files, op, Extension, OpState, ZeroCopyBuf};
+use libp2p::multiaddr::Protocol;
+use libp2p::{Multiaddr, PeerId};
+use peer::{PeerNode, PeerNodeConfig};
+
+mod peer;
+
+#[derive(Clone, Debug, Default)]
+pub struct Options {
+  /// Configuration options for the built-in (default) peer node
+  default_peer: PeerNodeConfig,
+}
+
+#[derive(Debug, Clone, Copy)]
+struct DefaultNodeResourceId(deno_core::ResourceId);
+
+pub fn init(options: Options) -> Extension {
+  Extension::builder(env!("CARGO_PKG_NAME"))
+    .js(include_js_files!(
+      prefix "internal:ext/libp2p",
+      "js/01_peer.js",
+    ))
+    .ops(vec![
+      op_p2p_get_peer_id::decl(),
+      op_p2p_request_protocol::decl(),
+    ])
+    .state(move |state| {
+      let default_node = PeerNode::spawn(options.default_peer.clone())
+        // FIXME: map errors to AnyError instead of panicking
+        // We need to convert `Box<dyn Error + Send>` to `anyhow::Error`
+        .unwrap();
+      let rid = state.resource_table.add(default_node);
+      state.put::<DefaultNodeResourceId>(DefaultNodeResourceId(rid));
+      Ok(())
+    })
+    .build()
+}
+
+#[op]
+pub fn op_p2p_get_peer_id(state: &mut OpState) -> Result<String> {
```

Comment:
is the `OpState` reference required to me mutable here? 🤔

Reply:
Yes, that's the way how Deno ops are designed. 

This is what happens when I change the reference to immutable:

```
error[E0277]: the trait bound `&OpState: Deserialize<'_>` is not satisfied
  --> ext/libp2p/lib.rs:44:1
   |
44 | #[op]
   | ^^^^^ the trait `Deserialize<'_>` is not implemented for `&OpState`
   |
   = help: the following other types implement trait `Deserialize<'de>`:
             &'a Path
             &'a [u8]
             &'a str
             ()
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
             (T0, T1, T2, T3, T4)
           and 193 others
note: required by a bound in `from_v8`
  --> /Users/bajtos/.cargo/registry/src/github.com-1ecc6299db9ec823/serde_v8-0.82.0/de.rs:48:6
   |
48 |   T: Deserialize<'de>,
   |      ^^^^^^^^^^^^^^^^ required by this bound in `from_v8`
   = note: this error originates in the attribute macro `op` (in Nightly builds, run with -Z macro-backtrace for more info)
```

The error could definitely be more helpful.

Reply:
Reported to Deno: 
- https://github.com/denoland/deno/issues/17894

Reply:
Makes sense, thank you!